### DISCUSSION TITLE
Loop with multiple targets

### DISCUSSION THREAD

**Discussion by bilogic** [livewire/livewire#4441] [7. Bugs (archived)]
> https://laravel-livewire.com/docs/2.x/loading-states

My interpretation of the docs is that:

1. We can specify multiple targets separated by commas
2. We can provide parameters for a target

However, it seems we cannot combine 1 and 2, i.e.

```
wire:target="saveItem(1)" // this works
wire:target="saveItem,increaseItemQuantity" // this also works
wire:target="saveItem(1),increaseItemQuantity(1)" // this throws an error "increaseItemQuantity is undefined"
```

I have a list of items generated by a for-loop
Each item has a numeric input field, `-`, `+` and a quantity field
I would like to do a `wire:loading.class` on the quantity field when its `-`/`+` is clicked or numeric field is updated.

**Comment by joshhanley** [comment]
@bilogic hmm yeah it seems it's not parsing the parameters when there is multiple targets.

It probably won't work, but can you see what happens if you put two `wire:targets` on the element, like `wire:target="saveItem(1)" wire:target="increaseItemQuantity(1)"`?

Thanks!

**Comment by bilogic** [comment]
@joshhanley 

Thanks let me try that out in a bit. I'm not familiar with the internals of livewire, but if multiple `wire:target` does not work, can I suggest considering `wire:targetRegex` or something along the lines?

The deeper challenge I faced was `wire:target` requires fixed targets, in other words, I cannot target `increaseItemQuantity(1, x)`, where there might be multiple values of `x` on the same page. 

So while I logically only needed to target 1 function `adjustItemQuantity(item_id, quantity)`, I now have to break them up into 3 functions: `increaseItemQuantity(item_id)`, `decreaseItemQuantity(item_id)` and `saveItem(item_id)`, the last of which requires alpine's `x-on:keyup.enter=` to run `setQuantity(item_id, quantity)` first, followed by `saveItem(item_id)`.

Hope that gives an idea of the challenge I'm facing. Thank you.

**Reply by joshhanley** [reply]
@bilogic yeah that's something Caleb would need to decide.

Hmm yeah makes sense. Where does quantity come from when passed into `adjustItemQuantity` or `setQuantity`?

**Reply by bilogic** [reply]
@joshhanley 

In `adjustItemQuantity`, it's `{{ blade_variable }} + 1` or `{{ blade_variable }} - 1` (for the `-` and `+` buttons)
In `setQuantity`, it comes from the HTML input tag, i.e. user types in a value and presses enter which triggers `x-on:keyup.enter`

In the ideal situation, I should only be using (and targeting) `adjustItemQuantity`, since its quantity is an absolute value.

Of course, I'm open to even better ways of coding this up. :)

**Reply by joshhanley** [reply]
@bilogic ah gotcha makes sense. Yep happy to have a look if you can share the code 🙂

**Reply by bilogic** [reply]
@joshhanley 

I don't have any code that works at the moment.

I was hoping this would work `wire:target="saveItem(1),increaseItemQuantity(1)" // this throws an error "increaseItemQuantity is undefined"` but it doesn't.
 
Ideally, I would love to have to only write `wire:targetRegex="adjustItemQuantity\({{ item_id }},.*\)"` => `wire:targetRegex="adjustItemQuantity\(1,.*\)"`.

**Reply by howdu** [reply]
@bilogic wondering if you ever found a solution for this?

I've come across the same limitation and `wire:targetRegex` would of been a lifesave.

**Reply by bilogic** [reply]
No, tbh, livewire is too blackbox for me

**Comment by rawaby88** [comment]
any solution to that problem yet? it 2 years already after

**Comment by classcruiser** [comment]
Still no solution it seems. I've recently tried two wire:target and it doesn't work. only the first wire:target appear on the element. so the solution for now is to have multiple loading component with different wire:target

**Reply by thehands79** [reply]
Is this not the solution?

https://github.com/livewire/livewire/discussions/5008

**Comment by achatzi78** [comment]
@thehands79 This solution does not work when you want multiple targets with parameters, for example `method1(123), method2(123), mthod3(123)`.

After exploring the js of Livewire 3, I found the source of the issue is in 2 places.

This part of code in livewire.js
```js
if (raw2.includes("(") && raw2.includes(")")) {
    targets.push({ target: directive3.method, params: quickHash(JSON.stringify(directive3.params)) });
}
```
calls this part of the code (this is a method in the `Directive` class)
```js
parseOutMethodAndParams(rawMethod) {
  let method = rawMethod;
  let params = [];
  const methodAndParamString = method.match(/(.*?)\((.*)\)/s);
  if (methodAndParamString) {
    method = methodAndParamString[1];
    let func = new Function("$event", `return (function () {
            for (var l=arguments.length, p=new Array(l), k=0; k<l; k++) {
                p[k] = arguments[k];
            }
            return [].concat(p);
        })(${methodAndParamString[2]})`);
    params = func(this.eventContext);
  }
  return { method, params };
}
```
which parses the string and assigns to method whatever is before the first parenthesis and to params whatever is between the first and last parenthesis.  So if we have a string like `methodName(123)` it will return `{methodName, [123]}`, but if we have a string like `methodName1(123), methodName2(123)` it will return `{methodName1, ['123), methodName2(123']}`.

This is why multiple targets with parameters do not work.

To resolve this, I added another getter to the `Directive` class called `methods`
```js
get methods() {
  const methodsAndParams = this.expression.matchAll(/\s*(?<functions>\w+)\((?<parameters>[^)]+)\)\s*(?:,|$)/g);
  let methods = [];

  for (const match of methodsAndParams) {
      let func = new Function("$event", `return (function () {
        for (var l=arguments.length, p=new Array(l), k=0; k<l; k++) {
            p[k] = arguments[k];
        }
        return [].concat(p);
    })(${match.groups.parameters})`);

      methods.push({ method: match.groups.functions, params: func(this.eventContext) });
  }

  return methods;
}
```
This returns an array of methods and their paramaters.  So if the target is `methodName1(123), methodName2(345)` it will return `[{methodName1, [123]}, {methodName2, [345]}]`.

 Then I also changed the following, from this
```js
if (raw2.includes("(") && raw2.includes(")")) {
    targets.push({ target: directive3.method, params: quickHash(JSON.stringify(directive3.params)) });
}
```
to this
```js
if (raw2.includes("(") && raw2.includes(")") && raw2.includes(",")) {
    for (const target of directive3.methods) {
        targets.push({ target: target.method, params: quickHash(JSON.stringify(target.params)) });
    }
  }
  else if (raw2.includes("(") && raw2.includes(")")) {
    targets.push({ target: directive3.method, params: quickHash(JSON.stringify(directive3.params)) });
  } 
```
so it first checks if there are parenthesis and commas and adds multiple targets.

I do not know if it affects something else (from my limited testing it seems to work) and my knowledge in javascript in very limited.  I am posting this here in case it helps someone or maybe the developer @calebporzio finds it useful to add in the future.

Best regards.

**Reply by blazery** [reply]
I would have loved this info when i started working on this xd. Maybe i should read the discussion more often. Im opening a Pr for this in a bit.
